### PR TITLE
feat: Make Event Saving, Draft Mode, Publishing Clearer 

### LINKED ISSUE #4673: Wizard: Make Event Saving, Draft Mode, Publishing Clearer
**1. When event is created and still unpublished (Draft Mode)**
* _Requirement_ for saving a draft is only
  1. to have an event name
* show buttons:
  * Discard (only when event was never saved - when creating the event: It does not save the event and discards any data that was entered. After that direct user back to dashboard of all events. If user has saved an event previously, there is no "Discard" possible anymore, then show "Cancel" and send user back to specific event dashboard.)
  * Cancel (cancels changes of the wizard step and exists to event dashboard)
  * Save/Previous (except for first step)
  * Save/Next (saves changes and goes to the next step, except for last step)
  * Save (saves draft and exits wizard to dashboard)
* below the above option show another section with two buttons
  * Show a horizontal line with 100% width of the page
  * Show the text "This event is currently not published. It is in draft mode and it is not visible publicly."
  * Preview (open in new tab)
  * Publish (opens a pop up - same as "Publish" on the dashboard / afterwards exits to event dashboard). Please show this button as "gray" as long as the event does not have the minimum data listed necessary for publishing an event.

Screenshot from 2020-07-30 18-16-43

**2. After an event is published already (Live Event)**
* _Requirement_ for an event to be publishable is, that it has
  1. A name
  2. A location or online event link
  3. A ticket
* below the above savings/next/previous options show another section with two buttons
  * Show the text "This event is published. Any changes you make will appear on your live event."
  * View (open in new tab)
  * Unpublish (opens a pop up - same as "Unpublish" on the dashboard / afterwards exits to event dashboard)

Screenshot from 2020-07-30 18-23-19

**3. Top menu "Wizard Steps"**

When a user clicks a top wizard menu item the following action should take place:
* Cancel changes of the current tab and load the menu tab

Screenshot from 2020-07-30 19-24-21

### PR DESCRIPTION
**Pull request by maze-runnar**
Fixes #4673

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by maze-runnar** [comment]
@iamareebjamal I have currently made changes for only first step wizard, if these changes are ok then I will proceed.

**Comment by maze-runnar** [comment]
adding preiew url on this page is giving error and blocking from publishing events.

**Comment by maze-runnar** [comment]
Screenshot from 2020-09-02 22-56-07
Screenshot from 2020-09-02 22-55-48

**Comment by maze-runnar** [comment]
adding cancel button without creating an event is not working, it can be added from the second wizard step or when user is not editing event first time.

**Comment by maze-runnar** [comment]
@iamareebjamal I need help with this issue, like what should be the approach, is my approach right or any other efficient way?
and also 

> adding cancel button without creating an event is not working, it can be added from the second wizard step or when user is not editing event first time.

**Comment by maze-runnar** [comment]
we can add a cancel and preview button once event is saved or published, but no way to know that event is saved at least once. can I add these buttons when `event.state == "published" `

**Comment by mariobehling** [comment]
> @iamareebjamal I need help with this issue, like what should be the approach, is my approach right or any other efficient way?
> and also
> 
> > adding cancel button without creating an event is not working, it can be added from the second wizard step or when user is not editing event first time.

Yes, that is expected. This is why the option in this case should be a "Discard" button. Eventbrite handles this in the same way. As long as an event is not "Saved as Draft" there can be no preview and it cannot be cancelled. Any information can only be discarded as long as the event is not saved.

**Comment by iamareebjamal** [comment]
Like I mentioned in the issue, `event.id` is null for unsaved events

**Comment by maze-runnar** [comment]
it's working now, thanks @iamareebjamal 
Screenshot from 2020-09-03 08-49-06
Screenshot from 2020-09-03 08-48-57
Screenshot from 2020-09-03 08-50-27

**Comment by mariobehling** [comment]
> Do i need to ad this in every wizard step, or it will be only in first step?

Discard is only in the first step as the wizard cannot progress if there is no location. Going through the details of the issue description again might make it clearer.

Please use the same spacing and design for the buttons as on the specific event dashboard.

Thanks

File: app/components/forms/wizard/basic-details-step.js [hunk 1]
```diff
@@ -83,6 +83,14 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return this.data.event.state !== 'published';
   }),
 
+  ticketNumber: computed('data.event.tickets', function() {
+    return this.data.event.tickets.length > 0 ? true : false;
+  }),
+
+  firstTimeEvent: computed('data.event.identifier', function() {
+    return (this.data.event.identifier === null || this.data.event.identifier=== "" || this.data.event.identifier=== undefined) ? true : false;
```
Comment (iamareebjamal) [inline app/components/forms/wizard/basic-details-step.js:91]:
```suggestion
    return !this.data.event.identifier; 
```

File: app/components/forms/wizard/basic-details-step.js [hunk 2]
```diff
@@ -83,6 +83,14 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return this.data.event.state !== 'published';
   }),
 
+  ticketNumber: computed('data.event.tickets', function() {
+    return this.data.event.tickets.length > 0 ? true : false;
+  }),
+
+  firstTimeEvent: computed('data.event.identifier', function() {
```
Comment (iamareebjamal) [inline app/components/forms/wizard/basic-details-step.js:90]:
```suggestion
  unsavedEvent: computed('data.event.identifier', function() {
```

File: app/components/forms/wizard/other-details-step.js [hunk 3]
```diff
@@ -39,6 +39,10 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return !!this.data.event.codeOfConduct;
   }),
 
+  ticketNumber: computed('data.event.tickets', function() {
+    return this.data.event.tickets.length > 0 ? true : false;
```
Comment (iamareebjamal) [inline app/components/forms/wizard/other-details-step.js:43]:
```suggestion
    return this.data.event.tickets.length;
```

File: app/components/forms/wizard/other-details-step.js [hunk 4]
```diff
@@ -39,6 +39,10 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return !!this.data.event.codeOfConduct;
   }),
 
+  ticketNumber: computed('data.event.tickets', function() {
```
Comment (iamareebjamal) [inline app/components/forms/wizard/other-details-step.js:42]:
```suggestion
  ticketsPresent: computed('data.event.tickets', function() {
```

File: app/components/forms/wizard/basic-details-step.js [hunk 5]
```diff
@@ -83,6 +83,14 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return this.data.event.state !== 'published';
   }),
 
+  ticketNumber: computed('data.event.tickets', function() {
+    return this.data.event.tickets.length > 0 ? true : false;
+  }),
+
+  unsavedEvent: computed('data.event.identifier', function() {
+    return !this.data.event.identifier;
```
Comment (niranjan94) [inline app/components/forms/wizard/basic-details-step.js:91]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/attendee-step.js [hunk 6]
```diff
@@ -12,6 +12,10 @@ export default Component.extend(FormMixin, {
     return this.data.customForms?.filter(field => !field.isFixed);
   }),
 
+  ticketNumber: computed('data.event.tickets', function() {
```
Comment (iamareebjamal) [inline app/components/forms/wizard/attendee-step.js:15]:
```suggestion
  ticketsPresent: computed('data.event.tickets.@each', function() {
```

**Comment by iamareebjamal** [comment]
@maze-runnar Please keep property names meaningful and avoid extra code. See reviews

File: app/templates/components/forms/wizard/basic-details-step.hbs [hunk 7]
```diff
@@ -617,7 +617,7 @@
     </div>
     <br><br><br>
     <div class="{{unless this.device.isMobile 'right floated large' 'mini three'}} ui fields {{if this.isUserUnverified 'disabled'}} buttons">
-    {{#if (not firstTimeEvent)}}
+    {{#if (not unsavedEvent)}}
```
Comment (iamareebjamal) [inline app/templates/components/forms/wizard/basic-details-step.hbs:620]:
```suggestion
    {{#if (not this.unsavedEvent)}}
```

Component properties without `this` is deprecated

**Comment by iamareebjamal** [comment]
If same functions are being used in all components, move them to a parent component and inherit from that

**Comment by maze-runnar** [comment]
> If same functions are being used in all components, move them to a parent component and inherit from that

actually, publish, save, and move are also used in every step and their function is written in every component, that's why I have written unpublish function at same place

**Comment by maze-runnar** [comment]
> @maze-runnar Please keep property names meaningful and avoid extra code. See reviews

ok, I am doing this

File: app/components/forms/wizard/sponsors-step.js [hunk 8]
```diff
@@ -70,6 +74,12 @@ export default Component.extend(FormMixin, {
       });
     }
   },
+    unpublish() {
+      this.onValid(() => {
+        this.set('data.event.state', 'draft');
+        this.sendAction('save', this.data);
+      });
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:81]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity decreasing per file
==============================
+ app/components/forms/wizard/attendee-step.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249&bid=20004936)

File: app/components/forms/wizard/sponsors-step.js [hunk 9]
```diff
@@ -70,6 +74,12 @@ export default Component.extend(FormMixin, {
...
+    },
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:82]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js [hunk 10]
```diff
@@ -70,6 +74,12 @@ export default Component.extend(FormMixin, {
       });
     }
   },
+    unpublish() {
+      this.onValid(() => {
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:78]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js [hunk 11]
```diff
@@ -70,6 +74,12 @@ export default Component.extend(FormMixin, {
       });
     }
   },
+    unpublish() {
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:77]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js [hunk 12]
```diff
@@ -70,6 +74,12 @@ export default Component.extend(FormMixin, {
...
+      this.onValid(() => {
+        this.set('data.event.state', 'draft');
+        this.sendAction('save', this.data);
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:80]: Codacy found an issue: [Expected indentation of 6 spaces but found 8.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js [hunk 13]
```diff
@@ -70,6 +74,12 @@ export default Component.extend(FormMixin, {
       });
     }
   },
+    unpublish() {
+      this.onValid(() => {
+        this.set('data.event.state', 'draft');
```
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:79]: Codacy found an issue: [Expected indentation of 6 spaces but found 8.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/basic-details-step.js (hunk 5 above)
Comment (niranjan94) [inline app/components/forms/wizard/basic-details-step.js:91]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js (hunk 8 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:81]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity decreasing per file
==============================
+ app/components/forms/wizard/attendee-step.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249&bid=20004936)

File: app/components/forms/wizard/sponsors-step.js (hunk 11 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:77]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/basic-details-step.js [hunk 14]
```diff
@@ -83,6 +83,14 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return this.data.event.state !== 'published';
   }),
 
+  ticketsPresent: computed('data.event.tickets.@each', function() {
+    return this.data.event.tickets.length > 0;
+  }),
+
+  unsavedEvent: computed('data.event.identifier', function() {
+    return !this.data.event.identifier;
```
Comment (niranjan94) [inline app/components/forms/wizard/basic-details-step.js:91]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js (hunk 13 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:79]: Codacy found an issue: [Expected indentation of 6 spaces but found 8.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js (hunk 9 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:82]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js (hunk 12 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:80]: Codacy found an issue: [Expected indentation of 6 spaces but found 8.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/sponsors-step.js (hunk 10 above)
Comment (niranjan94) [inline app/components/forms/wizard/sponsors-step.js:78]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

**Comment by iamareebjamal** [comment]
@maze-runnar Update?

File: app/components/forms/wizard/attendee-step.js [hunk 15]
```diff
@@ -12,6 +12,10 @@ export default Component.extend(FormMixin, {
     return this.data.customForms?.filter(field => !field.isFixed);
   }),
 
+  ticketsPresent: computed('data.event.tickets.@each', function() {
+    return this.data.event.tickets.length > 0;
+  }),
+
```
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:18]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/other-details-step.js [hunk 16]
```diff
@@ -39,6 +39,10 @@ export default Component.extend(FormMixin, EventWizardMixin, {
     return !!this.data.event.codeOfConduct;
   }),
 
+  ticketsPresent: computed('data.event.tickets.@each', function() {
+    return this.data.event.tickets.length > 0;
+  }),
```
Comment (iamareebjamal) [inline app/components/forms/wizard/other-details-step.js:44]: This is present in all 3 components

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/components/forms/wizard/sponsors-step.js  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249&bid=20004936)

File: app/components/forms/wizard/attendee-step.js (hunk 15 above)
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:18]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

File: app/components/forms/wizard/other-details-step.js (hunk 16 above)
Comment (maze-runnar) [reply app/components/forms/wizard/other-details-step.js:44]: can I put it in event-wizard? Will it work from there for all this components?

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249&bid=20004936)

File: app/components/forms/wizard/attendee-step.js (hunk 15 above)
Comment (niranjan94) [inline app/components/forms/wizard/attendee-step.js:18]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6112249)

**Comment by iamareebjamal** [comment]
@maze-runnar Your repo is deleted. I can't pull in changes to test

**Comment by maze-runnar** [comment]
> @maze-runnar Your repo is deleted. I can't pull in changes to test

can't it be tested on deployment URL please, or I have to create a new PR.

**Comment by iamareebjamal** [comment]
I have to make changes and I can't do that unless I pull te changes

**Comment by maze-runnar** [comment]
> I have to make changes and I can't do that unless I pull te changes

ok so I will create a new PR.

**Comment by maze-runnar** [comment]
go to https://github.com/fossasia/open-event-frontend/pull/5011

**Pull request closed without merging**